Added countdown to start of match.

LadyEbony opened the pull request:
### Motivation and Context
- [X] Why is this change required? What problem does it solve?
James asked me.
- [ ] If it fixes an open issue, include the text `Closes #1` (where 1 would be the issue number) to your commit message.

### Types of changes
What types of changes does your code introduce? Check all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
Added a countdown UI to Stage GUI.
Added boolean variable to MatchManager, which influences input from MatchController.
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Description
- [ ] Describe your changes in detail.
MatchManager now begins a countdown Task before the MatchStartEvent. During that period, new MatchManager variable, isControllerLocked, is set to true, which disables input from MatchController. Once the countdown finishes, isControllerLocked is set to False and MatchStartEvent is called as usual.

If I am doing something unpleasant in the code, please tell me. Not familiar with Tasks nor this project's code standards.

LadyEbony committed:
Initial comment to branch

LadyEbony committed:
Added countdown events.

LadyEbony committed:
Properly turned countdown to tasks. Added isControllerLocked

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownController.cs:
```diff
@@ -0,0 +1,45 @@
+public class MatchCountdownController : MonoBehaviour
+{
+    public Text TextUI;
+    public float GoDuration = 1.25f;
```
We are aiming to play some audio while this counts down. Potential option: take in a array of AudioClips, play them to an AudioSource and use that as the base for how long to wait. Should also avoid needing to hard-code the 3 second timer.

james7132 commented on Assets/Code/src/Runtime/Core/GameController.cs:
```diff
@@ -19,10 +19,12 @@ public class MatchController : IMatchController {
-    Assert.IsTrue(input.IsValid);
-    inputContext.Update(input);
+public virtual void Update() {
+    if (!MatchManager.Instance.isControllerLocked){
```
MatchController.Update is called mainly from MatchManager, where isControllerLocked is defined. I would advise just wrapping calls to it in MatchManager, not here.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownController.cs:
```diff
@@ -0,0 +1,45 @@
+
+public class MatchCountdownController : MonoBehaviour
+{
+    public Text TextUI;
```
We're trying to avoid use of normal Text objects when possible when we have TextMeshPro available. Please replace with a TextMeshProUGUI reference.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownController.cs:
```diff
@@ -0,0 +1,45 @@
+    /// Awake is called when the script instance is being loaded.
+    /// </summary>
+    void Awake()
+    {
```
Safety check: make sure that the text display object is not null.

james7132 reviewed (changes requested):
A few small changes, but otherwise a fast and quick implementation.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchManager.cs:
```diff
@@ -15,6 +15,7 @@ public class MatchManager : MonoBehaviour {
 
   public bool IsLocal => Config.IsLocal;
   public bool IsPaused { get; private set; }
+  public bool isControllerLocked { get; private set; }
```
Style: Please use uppercase properties for public properties. 

Might also be a good idea to also inverse this: IsControllerRunning, which defaults to false, set only in the RunMatch method.

LadyEbony committed:
Renamed MatchCountdownController to UI to match MatchEndUI name convention.
Changed Text to TextMeshProUGUI.
Added Audio countdown support.
Added intial delay to countdown.
Component references will search for refs when called and null.

LadyEbony committed:
Updated MatchManager's IsControllerRunning propetry to proper style.

LadyEbony commented:
Made changes requested. Check commits for all changes.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownUI.cs:
```diff
@@ -0,0 +1,107 @@
+    // If variables are null, look for them
+    [Header("Component References")]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public TextMeshProUGUI _TextUI;
```
Make this private, and annotate with `UnityEngine.SerializeField` instead. Same with below.

If it's private: `_textUI` vs `_TextUI`.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownUI.cs:
```diff
@@ -0,0 +1,107 @@
+    public TextMeshProUGUI _TextUI;
+    public TextMeshProUGUI TextUI {
+        get {
+            if (!_TextUI)
```
nit: Compare against null instead of using conversion to bool.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownUI.cs:
```diff
@@ -0,0 +1,107 @@
+    async Task StartCountdown(MatchStartCountdownEvent evt)
+    {
+        // For Debug purposes
+        if (DisableCountdown) return;
```
To make sure that this does not affect non-debug builds, boolean AND with `Debug.isDebugBuild`

james7132 commented on Assets/Code/src/Runtime/Match/Events/MatchStartCountdownEvent.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace HouraiTeahouse.FantasyCrescendo.Matches{
+
+public class MatchStartCountdownEvent : MatchEvent{
```
Quick doc comment here.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownUI.cs:
```diff
@@ -0,0 +1,107 @@
+        ObjectUtil.SetActive(TextUI.gameObject, true);
+
+        // TODO: Make timer value based on evt
+        var timer = 3;
```
Default this to the length of CountdownClips.

james7132 commented on Assets/Code/src/Runtime/Match/Components/MatchCountdownUI.cs:
```diff
@@ -0,0 +1,107 @@
+        var timer = 3;
+        while (timer > 0)
+        {
+            await UpdateTextUI(timer.ToString(), CountdownClips[timer], 1);
```
Please use the length of the clips as the wait length.

james7132 reviewed (changes requested):
Mostly small changes to make the code higher quality. Pretty much usable with the requested changes.

LadyEbony committed:
Made change requests to improve code quality.
Countdown is based on Audio clip length.

LadyEbony commented:
Made change requests to improve code quality.

james7132 reviewed (approved):
LGTM.
